cmd/daytona/config/ssh_file.go

```diff
@@ -104,7 +108,7 @@ func UnlinkSshFiles() error {
 
 // Add ssh entry
 
-func generateSshConfigEntry(profileId, workspaceId, projectName, knownHostsPath string) (string, error) {
+func generateSshConfigEntry(profileId, workspaceId, projectName, knownHostsPath string, gpgForward bool) (string, error) {
```

Comment:
![Screenshot 2024-10-11 at 09 57 10](https://github.com/user-attachments/assets/b34a9c03-e9bb-46b5-b823-0a3e66737fd3)

@bryans-go seems changes got messed up during the conflict resolution. This is the SSH config that gets generated for me know so GPG signing doesn't work anymore.

Please check the logic for generating the GPG signing ssh entry.

Reply:
![image](https://github.com/user-attachments/assets/6c6f5954-046a-4dbe-9669-d380d4df2cad)
Its working for me
![image](https://github.com/user-attachments/assets/9d91a1e8-702e-4ab4-bd6b-9dd6060cd2a7)

Reply:
I have tested GPG signing three to four times and it never hapened . maybe you the gitprovider you chose the wrong git provider which didn't had your key.

Reply:
also there were no conflicts in the file where I had written the gpg key entry generating function.

Reply:
@bryans-go 2 entries seem to be generated when you open the workspace in vscode. Please test this with `dtn code` for all IDEs.

Reply:
@Tpuljak I wanted to inform you that there are two entries because when we try to import the gpg key (which succeds) but when we export it in the workspace it will not let it export because the gpg key is already impoted from the local system.

Reply:
But we can not change the logic due the fact if user updates their signing key then it will not be updated accordingly.

Reply:
despite having 2 entries the gpg key is present on the system 
```Host default-2e793044fb36-todogolang
	User daytona
	StrictHostKeyChecking no
	UserKnownHostsFile /dev/null
	ProxyCommand "/home/bryan/Desktop/daytona/main" ssh-proxy default 2e793044fb36 todogolang
	ForwardAgent yes

Host default-2e793044fb36-todogolang
	User daytona
	StrictHostKeyChecking no
	UserKnownHostsFile /dev/null
	ProxyCommand "/home/bryan/Desktop/daytona/main" ssh-proxy default 2e793044fb36 todogolang
	ForwardAgent yes
	StreamLocalBindUnlink yes
	RemoteForward /home/daytona/.gnupg/S.gpg-agent:/run/user/1000/gnupg/S.gpg-agent.extra

```
this means that our main work was completed

Reply:
The config will get replaced into one config after the user updates the gpg key in their git provider settings

Reply:
> this means that our main work was completed

The main work is definitely not completed because it doesn't work 😄 

> @Tpuljak I wanted to inform you that there are two entries because when we try to import the gpg key (which succeds) but when we export it in the workspace it will not let it export because the gpg key is already impoted from the local system.

I'm not completely sure what this means but this solution **must** work every time the user wants to open the workspace. Please test the PR by opening the workspace with an IDE with `daytona code` and making sure that commits can be signed. Please post a video of you doing so so we can proceed to review.

Reply:
![image](https://github.com/user-attachments/assets/3e831795-e388-405d-a2b4-0459e80f6177)
@Tpuljak even on debugging it vigrously by printing the `gpgKey` in each and every corner of code and functions that edit / create / modify the ssh entry I'm not able to find a cause for it but as I said that our main motive of exporting the public GPGkey got success and hence the git commits are GPG signed

Reply:
![Screenshot_20241011_173303](https://github.com/user-attachments/assets/81bfdaa1-e586-4c96-8edb-80f88063db11)
![Screenshot_20241011_173336](https://github.com/user-attachments/assets/b00fe168-5c5e-4550-aea4-6950f7c8b67f)

Reply:
@bryans-go the GPG is imported but you are not able to sign because the agent is not forwarded inside the workspace.

Please record a video of the following:
1. daytona create
2. daytona code (with vscode)
3. modify something
4. git commit -S -m "test commit"
5. git log --show-signature

Reply:
as I said earlier I'm **able** sign commits with GPG

https://github.com/user-attachments/assets/690b151e-8098-4f3b-a5ce-e4e8ef77ee91

Reply:
Do you have `StreamLocalBindUplink` and `RemoteForward` set in your `~/.ssh/config` (not `~/.ssh/daytona_config`)?

Reply:
![image](https://github.com/user-attachments/assets/ee639ffb-6970-4a1a-922a-3e5ca7f5f20f)
no

Reply:
My ssh config only has the following line as seen in the image

Reply:
As I said earlier our main motive is to export the gpg key to remote machine then it will handle all things
https://github.com/daytonaio/daytona/pull/1146#discussion_r1796710153

Reply:
Figured out the issue.

2 modifications will be required:
1. `ssh_file.go` replace the existing 2 functions with these 2

```golang
func EnsureSshConfigEntryAdded(profileId, workspaceName, projectName string, gpgKey string) error {
	err := ensureSshFilesLinked()
	if err != nil {
		return err
	}

	sshDir := filepath.Join(sshHomeDir, ".ssh")
	configPath := filepath.Join(sshDir, "daytona_config")

	knownHostsFile := getKnownHostsFile()

	// Read existing content from the file
	existingContent, err := os.ReadFile(configPath)
	if err != nil && !os.IsNotExist(err) {
		return err
	}

	// Generate SSH config entry without GPG forwarding
	newContent, err := appendSshConfigEntry(configPath, profileId, workspaceName, projectName, knownHostsFile, false, string(existingContent))
	if err != nil {
		return err
	}

	if gpgKey != "" {
		// Generate SSH config entry with GPG forwarding and override previous config
		_, err := appendSshConfigEntry(configPath, profileId, workspaceName, projectName, knownHostsFile, true, newContent)
		if err != nil {
			return err
		}
		projectHostname := GetProjectHostname(profileId, workspaceName, projectName)
		err = ExportGPGKey(gpgKey, projectHostname)
		if err != nil {
			return err
		}
	}

	return nil
}

func appendSshConfigEntry(configPath, profileId, workspaceId, projectName, knownHostsFile string, gpgForward bool, existingContent string) (string, error) {
	data, err := generateSshConfigEntry(profileId, workspaceId, projectName, knownHostsFile, gpgForward)
	if err != nil {
		return "", err
	}

	if strings.Contains(existingContent, data) {
		// Entry already exists in the file
		return existingContent, nil
	}

	// We want to remove the config entry gpg counterpart
	configCounterpart, err := generateSshConfigEntry(profileId, workspaceId, projectName, knownHostsFile, !gpgForward)
	if err != nil {
		return "", err
	}
	updatedContent := strings.ReplaceAll(existingContent, configCounterpart, "")
	updatedContent = data + updatedContent

	// Open the file for writing
	file, err := os.Create(configPath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	_, err = file.WriteString(updatedContent)
	return updatedContent, err
}
```

2. `pkg/cmd/workspace/create.go`

Place this above the `waitForDial` call and remove the gpgKey getting below (before openIDE)
```golang
		gpgKey, err := GetGitProviderGpgKey(apiClient, ctx, projects[0].Source.Repository.Url)
		if err != nil {
			log.Warn(err)
		}

```

Now the ssh entries are correctly generated and only one of them is present at any point

Reply:
pushed changes
![image](https://github.com/user-attachments/assets/05157fec-b18e-4d24-8dc1-3dd171238c66)
now working perfectly

Reply:
@Tpuljak thanks a lot.
for this great help